pkg/kube/watcher/watcher.go

```diff
@@ -0,0 +1,94 @@
+package watcher
+
+import (
+	"fmt"
+
+	"github.com/jenkins-x/jx/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// Watcher lists and watches resources
+type Watcher struct {
+	Namespace    string
+	ListOptions  metav1.ListOptions
+	HandlerFuncs HandlerFuncs
+	ListWatch    ListWatch
+}
+
+// HandlerFuncs is an adaptor to let you easily specify as many or
+// as few of the notification functions as you want
+type HandlerFuncs struct {
+	AddFunc    func(obj runtime.Object)
+	ModifyFunc func(obj runtime.Object)
+	DeleteFunc func(obj runtime.Object)
+}
+
+// ListWatch an interface for something that can list and watch resources
+type ListWatch interface {
+	List(ns string, listOptions metav1.ListOptions) ([]runtime.Object, error)
+	Watch(ns string, listOptions metav1.ListOptions) (watch.Interface, error)
+}
+
+// Run watches the resources until the connection is lost
+func (w *Watcher) Run() error {
```

Comment:
It might be helpful to have a graceful termination for the watcher. An `ctx context.Context` can be defined as an argument of `Run` and then in the select the termination can be handled:
```Go
select {
    case <-ctx.Done():
	watcher.Stop()
	return
}
```